### PR TITLE
Automated cherry pick of #128305

### PR DESCRIPTION
**Pull request by adrianmoisey**
Cherry pick of #128305 on release-1.31.

#128305: Ensure that a node's CIDR isn't released until the node is

```release-note
Fix bug where PodCIDR was released before node was deleted.
```

### CODE REVIEW

**Comment by adrianmoisey** [comment]
> If you **didn't cherry-pick** this change to [**all supported release branches**](https://k8s.io/releases/patch-releases), please leave a comment describing why other cherry-picks are not needed to speed up the review process.

This bug was first discovered in Kubernetes 1.31, so only cherry-picked into that branch.

> Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

I'm unsure on the correct procedure to fix this, considering it's a cherry-pick

**Comment by aojea** [comment]
/lgtm
/approve

linter errcheck error we omitted in the master branch :/  

The next assertions on the test will fail if there is an error, but we should validate all errors, it seems now is too late

https://github.com/kubernetes/kubernetes/pull/128305/

**Comment by aojea** [comment]
/kind bug
@adrianmoisey it needs a release note

**Comment by adrianmoisey** [comment]
/release-note-edit ```release-note Fix bug where PodCIDR was released before node was deleted```

**Comment by adrianmoisey** [comment]
/release-note-edit ```release-note Fix bug where PodCIDR was released before node was deleted```

File: pkg/controller/nodeipam/ipam/range_allocator_test.go [hunk 1]
```diff
@@ -841,3 +841,109 @@ func TestReleaseCIDRSuccess(t *testing.T) {
...
+				t.Fatalf("failed to create CIDRRangeAllocator: %v", err)
+			}
+			rangeAllocator, ok := allocator.(*rangeAllocator)
+			if !ok {
+				t.Fatalf("found non-default implementation of CIDRAllocator")
+			}
+			rangeAllocator.nodesSynced = test.AlwaysReady
+			rangeAllocator.recorder = testutil.NewFakeRecorder()
+
+			rangeAllocator.syncNode(tCtx, tc.nodeKey)
```
Comment (saschagrunert) [inline pkg/controller/nodeipam/ipam/range_allocator_test.go:929]: This seems legit:

```
ERROR: pkg/controller/nodeipam/ipam/range_allocator_test.go:929:27: Error return value of `rangeAllocator.syncNode` is not checked (errcheck)
ERROR: 			rangeAllocator.syncNode(tCtx, tc.nodeKey)
```

**Comment by adrianmoisey** [comment]
Fixed this over in https://github.com/kubernetes/kubernetes/pull/128806

**Pull request closed without merging**